confluent_kafka/avro/serializer/message_serializer.py

```diff
@@ -155,34 +161,36 @@ def _get_decoder_func(self, schema_id, payload):
         if schema_id in self.id_to_decoder_func:
             return self.id_to_decoder_func[schema_id]
 
-        # fetch from schema reg
-        try:
-            schema = self.registry_client.get_by_id(schema_id)
-        except:
-            schema = None
-
-        if not schema:
-            err = "unable to fetch schema with id %d" % (schema_id)
-            raise SerializerError(err)
-
         curr_pos = payload.tell()
         if HAS_FAST:
             # try to use fast avro
             try:
-                schema_dict = schema.to_json()
-                read_data(payload, schema_dict)
+                schema = self.registry_client.get_by_id(schema_id, json_format=True)
+                # here means we passed so this is something fastavro can do
+                # seek back since it will be called again for the
+                # same payload - one time hit
+                read_data(payload, schema)
 
                 # If we reach this point, this means we have fastavro and it can
                 # do this deserialization. Rewind since this method just determines
                 # the reader function and we need to deserialize again along the
                 # normal path.
                 payload.seek(curr_pos)
-
-                self.id_to_decoder_func[schema_id] = lambda p: read_data(p, schema_dict)
+                self.id_to_decoder_func[schema_id] = lambda p: read_data(p, schema)
                 return self.id_to_decoder_func[schema_id]
             except:
                 pass
 
+        # fetch from schema reg
+        try:
+            schema = self.registry_client.get_by_id(schema_id)
+        except:
+            schema = None
+
+        if not schema:
```

Comment:
this was just moved - used to be before attempting to use fastavro, but I can change it